animation_nodes/nodes/color/set_vertex_color.py

```diff
@@ -1,62 +1,129 @@
+
+        if self.colorLayerIdentifierType == "INDEX":
+            self.newInput("Integer", "Color Index", "colorLayerIndex")
+        elif self.colorLayerIdentifierType == "NAME":
+            self.newInput("Text", "Name", "colorLayerName")
+
+        self.newInput(VectorizedSocket("Color", "useColorList",
+            ("Color", "color"), ("Colors", "colors")))
+
         self.newOutput("Object", "Object", "outObject")
 
     def draw(self, layout):
-        layout.prop(self, "vertexColorName", text = "", icon = "GROUP_VCOL")
+        layout.prop(self, "mode", text = "")
 
     def drawAdvanced(self, layout):
-        layout.prop(self, "checkIfColorIsSet", text = "Check Color")
+        layout.prop(self, "colorLayerIdentifierType", text = "Type")
 
-    def execute(self, object, color):
-        if object is None: return object
-        if object.type != "MESH": return object
-        if object.mode == "EDIT":
-            self.raiseErrorMessage("Object is in edit mode")
-            return object
+    def getExecutionFunctionName(self):
+        if self.useColorList:
+            return "execute_ColorsList"
+        else:
+            return "execute_SingleColor"
+
+    def execute_SingleColor(self, object, identifier, color):
+        if object is None: return
+        colorLayer = self.getVertexColorLayer(object, identifier)
 
-        mesh = object.data
+        colorsList = VirtualColorList.create(color, 0).materialize(len(colorLayer.data))
 
-        # Vertex Colors are internally stored with 8 bytes
-        # I compress the color already here for an easier comparison
-        newColor = tuple(min(max(int(x * 255) / 255, 0), 1) for x in color)
+        colorLayer.data.foreach_set("color", colorsList.asNumpyArray())
+        object.data.update()
+        return object  
+        
+    def execute_ColorsList(self, object, identifier, colors):
+        if object is None: return
 
-        colorLayer = getVertexColorLayer(mesh, self.vertexColorName)
-        if len(colorLayer.data) == 0: return object
+        if self.mode == "LOOPS":
+            return self.setVertexColorForLoops(object, identifier, colors) 
 
-        if self.checkIfColorIsSet:
-            oldColor = colorLayer.data[0].color
-            if colorsAreEqual(newColor, oldColor):
-                return object
+        elif self.mode == "VERTICES":
+            return self.setVertexColorForVertices(object, identifier, colors)
 
-        newColorAsArray = numpy.array(newColor, dtype = "f")
-        colors = numpy.tile(newColorAsArray, len(colorLayer.data))
-        colorLayer.data.foreach_set("color", colors)
+        elif self.mode == "POLYGONS":
+            return self.setVertexColorForPolygons(object, identifier, colors)
 
-        # without this line the 3d view doesn't update
-        colorLayer.data[0].color = newColor
+    def setVertexColorForLoops(self, object, identifier, colors):
```

Comment:
You have a lot of repeated code here. I would say just implement them in a single function.

Reply:
Do you mean something like this,
```
def execute_ColorsList(self, object, identifier, colors):
        if object is None: return

        colorLayer = self.getVertexColorLayer(object, identifier)
        
        if self.mode == "LOOPS":
            colorsList = VirtualColorList.create(colors, 0).materialize(len(colorLayer.data))
            
        elif self.mode == "VERTICES":
            sourceMesh = object.an.getMesh(False)
            polygonIndices = sourceMesh.an.getPolygonIndices()
        
            verticesColors = VirtualColorList.create(colors, 0)
            colorsList = getLoopColorsFromVertexColors(polygonIndices, verticesColors) 

        elif self.mode == "POLYGONS":
            sourceMesh = object.an.getMesh(False)
            polygonIndices = sourceMesh.an.getPolygonIndices()
        
            polygonsColors = VirtualColorList.create(colors, 0)
            colorsList = getLoopColorsFromPolygonColors(polygonIndices, polygonsColors)
        
        colorLayer.data.foreach_set("color", colorsList.asNumpyArray())
        object.data.update()
        return object
```

Reply:
Yes, but you can also move the VirtualColorList creation outside of the if statement.

Reply:
Now, it looks like this,
```
def execute_ColorsList(self, object, identifier, colors):
        if object is None: return
        
        defaultColor = Color((0, 0, 0, 1))
        colorsList = VirtualColorList.create(colors, defaultColor)
        colorLayer = self.getVertexColorLayer(object, identifier)
        
        if self.mode == "LOOP":
            colorsList = colorsList.materialize(len(colorLayer.data))
            
        elif self.mode == "VERTEX":
            sourceMesh = object.an.getMesh(False)
            polygonIndices = sourceMesh.an.getPolygonIndices()

            colorsList = getLoopColorsFromVertexColors(polygonIndices, colorsList) 

        elif self.mode == "POLYGON":
            sourceMesh = object.an.getMesh(False)
            polygonIndices = sourceMesh.an.getPolygonIndices()
        
            colorsList = getLoopColorsFromPolygonColors(polygonIndices, colorsList)
        
        colorLayer.data.foreach_set("color", colorsList.asNumpyArray())
        object.data.update()
        return object
```

Reply:
I have also changed the name of modes,
Loops -> Loop
Vertices -> Vertex
Polygons -> Polygon

Reply:
Looks good, but maybe remove the newline before the elifs.